saliency/xrai.py

```diff
@@ -0,0 +1,533 @@
+  return im
+
+
+def _get_segments_felsenschwab(im,
+                               resize_image=True,
+                               scale_range=None,
+                               dilation_rad=5):
+  """Compute image segments based on felsenschwab algorithm
+
+  Args:
+    im: Input image.
+    resize_image: If resize_image is True, images are resized to 224,224 for
+                  segmentation output segments are always the same size as a
+                  the input image. This is for consistency w.r.t. segmentation
+                  parameter range. Defaults to True.
+    scale_range: Range of image values to use for segmentation algorithm.
+                  Segmentation algorithm is sensitive to the input image
+                  values, therefore we need to be consistent with the range
+                  for all images. Defaults to None.
+    dilation_rad: Sets how much each segment is dilated to include edges,
+                  larger values cause more blobby segments, smaller values
+                  get sharper areas. Defaults to 5.
+  Returns:
+      masks: List of np.ndarrays size of HxW for im size of HxWxC
+  """
+
+  # TODO (tolgab) Set this to default float range of 0.0 - 1.0 and tune
+  # parameters for that
+  if scale_range is None:
+    scale_range = [-1.0, 1.0]
+  SCALE_VALUES = [50, 100, 150, 250, 500, 1200]
+  SIGMA_VALUES = [0.8]
+  # Normalize image value range and size
+  original_shape = im.shape[:2]
+  # TODO (tolgab) This resize is unnecessary with more intelligent param range
+  # selection
+  if resize_image:
+    im = _normalize_image(im, scale_range, (224, 224))
+  else:
+    im = _normalize_image(im, scale_range)
+  segs = []
+  for scale in SCALE_VALUES:
+    for sigma in SIGMA_VALUES:
+      seg = segmentation.felzenszwalb(im, scale=scale, sigma=sigma, min_size=200)
+      if resize_image:
+        seg = resize(seg,
+                     original_shape,
+                     order=0,
+                     preserve_range=True,
+                     mode='constant',
+                     anti_aliasing=False).astype(np.int)
+      segs.append(seg)
+  masks = _unpack_segs_to_masks(segs)
+  if dilation_rad:
+    selem = disk(dilation_rad)
+    masks = [dilation(mask, selem=selem) for mask in masks]
+  return masks
+
+
+def _attr_aggregation_max(attr, axis=-1):
+  return attr.max(axis=axis)
+
+
+def _gain_density(mask1, attr, mask2=None):
+  # Compute the attr density over mask1. If mask2 is specified, compute density
+  # for mask1 \ mask2
+  if mask2 is None:
+    added_mask = mask1
+  else:
+    added_mask = _get_diff_mask(mask1, mask2)
+  return attr[added_mask].mean()
+
+
+def _get_diff_mask(add_mask, base_mask):
+  return np.logical_and(add_mask, np.logical_not(base_mask))
+
+
+def _get_diff_cnt(add_mask, base_mask):
+  return np.sum(_get_diff_mask(add_mask, base_mask))
+
+
+def _unpack_segs_to_masks(segs):
+  masks = []
+  for seg in segs:
+    for l in xrange(seg.min(), seg.max() + 1):
```

Comment:
'xrange' is not defined in Python 3. Use 'range()' or other alternative.

Reply:
done, although in future we may use six.